tuna.go

```diff
@@ -494,59 +494,83 @@ func (c *Common) CreateServerConn(force bool) error {
+	costChan := make(chan common.Fixed64)
+	resetChan := make(chan struct{})
 
-	time.Sleep(DefaultNanoPayUpdateInterval)
+	go func() {
+		for {
+			time.Sleep(time.Second)
+			bytesIn := atomic.LoadUint64(bytesInUsed)
+			bytesOut := atomic.LoadUint64(bytesOutUsed)
+			if (bytesIn+bytesOut)-(*bytesInPaid+*bytesOutPaid) >= uint64(minChargingUnit*TrafficUnit) {
+				cost := exitToEntryPrice*common.Fixed64(bytesIn-*bytesInPaid)/TrafficUnit + entryToExitPrice*common.Fixed64(bytesOut-*bytesOutPaid)/TrafficUnit
+				*bytesInPaid = bytesIn
+				*bytesOutPaid = bytesOut
+				lastCost = cost
+				costChan <- cost
+			}
+		}
+	}()
+
+	go func() {
+	loop:
+		for {
+			select {
+			case <-resetChan:
+				goto loop
+			case <-time.After(DefaultNanoPayUpdateInterval):
+				bytesIn := atomic.LoadUint64(bytesInUsed)
+				bytesOut := atomic.LoadUint64(bytesOutUsed)
+				cost := exitToEntryPrice*common.Fixed64(bytesIn-*bytesInPaid)/TrafficUnit + entryToExitPrice*common.Fixed64(bytesOut-*bytesOutPaid)/TrafficUnit
+				if cost == lastCost {
+					<-time.After(DefaultNanoPayUpdateInterval / 10)
+					break
+				}
+				*bytesInPaid = bytesIn
+				*bytesOutPaid = bytesOut
+				lastCost = cost
+				costChan <- cost
+			}
+		}
+	}()
 
 	for {
 		select {
 		case <-closeChan:
 			return
-		default:
-		}
-
-		bytesIn := atomic.LoadUint64(bytesInUsed)
-		bytesOut := atomic.LoadUint64(bytesOutUsed)
-		cost := exitToEntryPrice*common.Fixed64(bytesIn-*bytesInPaid)/TrafficUnit + entryToExitPrice*common.Fixed64(bytesOut-*bytesOutPaid)/TrafficUnit
-		if cost == lastCost {
-			time.Sleep(DefaultNanoPayUpdateInterval / 10)
-			continue
-		}
-		lastCost = cost
+		case cost := <-costChan:
+			paymentStream, err := getPaymentStream()
+			if err != nil {
+				log.Printf("get payment stream err: %v", err)
+				time.Sleep(time.Second)
+				continue
+			}
 
-		paymentStream, err := getPaymentStream()
-		if err != nil {
-			log.Printf("get payment stream err: %v", err)
-			time.Sleep(time.Second)
-			continue
-		}
+			paymentReceiver := c.GetPaymentReceiver()
+			if np == nil || np.Recipient() != paymentReceiver {
+				np, err = c.Wallet.NewNanoPay(paymentReceiver, nanoPayFee, DefaultNanoPayDuration)
+				if err != nil {
+					log.Printf("create nano payment err: %v", err)
+					time.Sleep(time.Second)
+					continue
+				}
+			}
 
-		paymentReceiver := c.GetPaymentReceiver()
-		if np == nil || np.Recipient() != paymentReceiver {
-			np, err = c.Wallet.NewNanoPay(paymentReceiver, nanoPayFee, DefaultNanoPayDuration)
+			err = sendNanoPay(np, paymentStream, cost)
 			if err != nil {
-				log.Printf("create nano payment err: %v", err)
+				log.Printf("send nano payment err: %v", err)
 				time.Sleep(time.Second)
 				continue
 			}
+			resetChan <- struct{}{}
 		}
-
-		err = sendNanoPay(np, paymentStream, cost)
-		if err != nil {
-			log.Printf("send nano payment err: %v", err)
-			time.Sleep(time.Second)
-			continue
-		}
-
-		*bytesInPaid = bytesIn
```

Comment:
You shouldn't remove this. `bytesInPaid` and `bytesOutPaid` should ONLY be updated if send nanopay success